### PR TITLE
rbd: remove unnecessary C wrapper functions

### PR DESCRIPTION
**Pull request by ansiwen**
Signed-off-by: Sven Anderson <sven@redhat.com>

## Checklist
- [x] Added tests for features and functional changes
- [x] Public functions and types are documented
- [x] Standard formatting is applied to Go code

### CODE REVIEW

**Review by nixpanic — APPROVED** [review]

File: internal/cutil/cutil.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package cutil
+
+import "unsafe"
+
+// VoidPtr casts a uintptr value to an unsafe.Pointer value in order to use it
+// directly as a void* argument in a C function call.
+// CAUTION: NEVER store the result in a variable, or the Go GC could panic.
+func VoidPtr(i uintptr) unsafe.Pointer {
+	var nullPtr unsafe.Pointer
+	return unsafe.Pointer(uintptr(nullPtr) + i)
```
Comment (phlogistonjohn) [inline internal/cutil/cutil.go:10]: What's this nullPtr do? Is it to avoid go vet errors?

Comment (ansiwen) [reply internal/cutil/cutil.go:10]: Yes, you can only cast a `uintptr` to `unsafe.Pointer` if you have cast it from `unsafe.Pointer` to `uintptr` in the same expression for pointer arithmetics.

**Comment by ansiwen** [comment]
> ```
> void * void_ptr(uintptr_t v) {
>     return (void*)v;
> }
> ```
> 
> Would that work with less risk?

Ha, it might be even easier than that! Let my try something....

**Comment by ansiwen** [comment]
Unfortunately, what I had in mind, that is a `typedef void_ptr (void *);` and then using a cast like `C.void_ptr(...)` didn't work. But I agree with you, it's safer to use a C function for that. It was not so easy to implement that only at a single place for the whole project. Now I just put everything in a single C header file. PTAL.

**Review by nixpanic — APPROVED** [review]

**Comment by ansiwen** [comment]
@phlogistonjohn Ok, after thinking about it a bit more: is it really less risky? Of course one could still store the result of `C.void_ptr(10)` in a variable and would have exactly the same problem with the GC. We just obfuscated the same function a bit, by moving it into the C space. So now, and also because the header file solution is not so pretty, I think I prefer the Go function with a clear warning again. 😅
WDYT?

**Comment by phlogistonjohn** [comment]
> @phlogistonjohn Ok, after thinking about it a bit more: is it really less risky? Of course one could still store the result of `C.void_ptr(10)` in a variable and would have exactly the same problem with the GC. 

Would it? Honest question, not rhetorical. I would have thought that keeping it "outside" the go memory realm the GC woudln't touch it. If I'm wrong about this that certainly changes my opinion.

> We just obfuscated the same function a bit, by moving it into the C space. So now, and also because the header file solution is not so pretty, I think I prefer the Go function with a clear warning again. sweat_smile
> WDYT?

Yeah, if the GC will touch it and there are no advantages around reducing the risk of a crash I agree with this and would prefer the earlier approach.

**Comment by ansiwen** [comment]
> > @phlogistonjohn Ok, after thinking about it a bit more: is it really less risky? Of course one could still store the result of `C.void_ptr(10)` in a variable and would have exactly the same problem with the GC.
> 
> Would it? Honest question, not rhetorical. I would have thought that keeping it "outside" the go memory realm the GC woudln't touch it. If I'm wrong about this that certainly changes my opinion.

Yeah, because by storing the value in an `unsafe.Pointer` variable, it's not "outside" of go memory realm anymore. The GC will check it occasionally and will bail out with something like
```
runtime: bad pointer in frame main.main.func1 at 0xc000050fb0: 0x1e4
fatal error: invalid pointer found on stack
```
I wrote reproducers for both variants, try for yourself: https://gist.github.com/ansiwen/77c1d53ff17bbc6c53424f17c1b59090

> 
> > We just obfuscated the same function a bit, by moving it into the C space. So now, and also because the header file solution is not so pretty, I think I prefer the Go function with a clear warning again. sweat_smile
> > WDYT?
> 
> Yeah, if the GC will touch it and there are no advantages around reducing the risk of a crash I agree with this and would prefer the earlier approach.

Not "touch" it, but look at it and become unhappy. Which is enough to crash. ;-)

**Comment by phlogistonjohn** [comment]
> > > @phlogistonjohn Ok, after thinking about it a bit more: is it really less risky? Of course one could still store the result of `C.void_ptr(10)` in a variable and would have exactly the same problem with the GC.
> > 
> > 
> > Would it? Honest question, not rhetorical. I would have thought that keeping it "outside" the go memory realm the GC woudln't touch it. If I'm wrong about this that certainly changes my opinion.
> 
> Yeah, because by storing the value in an `unsafe.Pointer` variable, it's not "outside" of go memory realm anymore. The GC will check it occasionally and will bail out with something like
> 
> ```
> runtime: bad pointer in frame main.main.func1 at 0xc000050fb0: 0x1e4
> fatal error: invalid pointer found on stack
> ```
> 

Thanks for that. It's unfortunate, but based on this I'd rather go with your original Go approach then. I think that's easier to maintain & work with. The extra pointer arithmetic thing done just to avoid go vet is unfortunate, but I can live with it. Could you add a comment to that function to serve as a reminder that's the reason for the extra variable and math, please?


> I wrote reproducers for both variants, try for yourself: https://gist.github.com/ansiwen/77c1d53ff17bbc6c53424f17c1b59090
> 
> > > We just obfuscated the same function a bit, by moving it into the C space. So now, and also because the header file solution is not so pretty, I think I prefer the Go function with a clear warning again. sweat_smile
> > > WDYT?
> > 
> > 
> > Yeah, if the GC will touch it and there are no advantages around reducing the risk of a crash I agree with this and would prefer the earlier approach.
> 
> Not "touch" it, but look at it and become unhappy. Which is enough to crash. ;-)

Oh, that's too bad. 

Longer term I wonder if this is worth communicating with the go team about. This is the first thing in cgo and unsafe packages that feels actually "broken" to me.

**Comment by phlogistonjohn** [comment]
One last stray thought: what if we passed the index via a pointer to the index value. This would cause the unsafe.Pointer value to actually be a pointer value. It's a tad inefficient IMO but if it's less risky it might be better. Or would that violate the pointer passing rules?

**Comment by ansiwen** [comment]
> One last stray thought: what if we passed the index via a pointer to the index value. This would cause the unsafe.Pointer value to actually be a pointer value. It's a tad inefficient IMO but if it's less risky it might be better. Or would that violate the pointer passing rules?

@phlogistonjohn Yes, AFAIK that would violate the rules. If the memory has been allocated by the Go memory management, pointers are not necessarily constant. The GC might move the memory and adjust all pointers referencing it. If you pass such a pointer into C code, it might become invalid. A way to avoid that is to allocate memory with malloc, which will not be touched by the GC. And that is exactly what https://github.com/mattn/go-pointer does, it allocates 1 byte for each pointer it needs. So if we want to go that direction, we should just use that package, I guess.

**Comment by ansiwen** [comment]
> LGTM Thanks! Its a really nice reduction of boilerplate as long as we use the new helper function with proper respect. :-)

@phlogistonjohn: BTW, the same respect we have to show whenever C calls a Go callback with a `void*`/`unsafe.Pointer` as argument, because that could also be a dangling pointer, so storing that as an unsafe.Pointer somewhere would also potentially crash. Whenever we receive a `void*` from C that might be dangling, the resulting `unsafe.Pointer` must be immediately converted to `uintptr` before storing it. Again, the only way to avoid this, is letting C allocate the memory for these pointers.

**Pull request merged**